update/sync/syncTransactions.go

```diff
@@ -0,0 +1,263 @@
+	}
+	p.mutPendingTx.Unlock()
+
+	var err error
+	defer func() {
+		p.mutPendingTx.Lock()
+		p.stopSync = true
+		if err == nil {
+			p.syncedAll = true
+		}
+		p.mutPendingTx.Unlock()
+	}()
+
+	if requestedTxs > 0 {
+		err = WaitFor(p.chReceivedAll, waitTime)
+		if err != nil {
+			log.Warn("could not finish syncing", "error", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *pendingTransactions) requestTransactionsFor(miniBlock *block.MiniBlock) int {
+	p.mutPendingTx.Lock()
+	defer p.mutPendingTx.Unlock()
+
+	missingTxs := make([][]byte, 0)
+	for _, txHash := range miniBlock.TxHashes {
+		if _, ok := p.mapTransactions[string(txHash)]; ok {
+			continue
+		}
+
+		tx, ok := p.getTransactionFromPoolOrStorage(txHash)
+		if ok {
+			p.mapTransactions[string(txHash)] = tx
+			continue
+		}
+
+		missingTxs = append(missingTxs, txHash)
+	}
+
+	for _, txHash := range missingTxs {
+		p.mapHashes[string(txHash)] = miniBlock
+	}
+
+	switch miniBlock.Type {
+	case block.TxBlock:
+		p.requestHandler.RequestTransaction(miniBlock.SenderShardID, missingTxs)
+	case block.SmartContractResultBlock:
+		p.requestHandler.RequestUnsignedTransactions(miniBlock.SenderShardID, missingTxs)
+	case block.RewardsBlock:
+		p.requestHandler.RequestRewardTransactions(miniBlock.SenderShardID, missingTxs)
+	}
+
+	return len(missingTxs)
+}
+
+// receivedMiniBlock is a callback function when a new transactions was received
+func (p *pendingTransactions) receivedTransaction(txHash []byte) {
+	p.mutPendingTx.Lock()
+	if p.stopSync {
+		p.mutPendingTx.Unlock()
+		return
+	}
+
+	if _, ok := p.mapHashes[string(txHash)]; ok {
+		p.mutPendingTx.Unlock()
+		return
+	}
+	if _, ok := p.mapTransactions[string(txHash)]; ok {
+		p.mutPendingTx.Unlock()
+		return
+	}
+
+	tx, ok := p.getTransactionFromPool(txHash)
+	if !ok {
+		p.mutPendingTx.Unlock()
+		return
+	}
+
+	p.mapTransactions[string(txHash)] = tx
+	receivedAllMissing := len(p.mapHashes) == len(p.mapTransactions)
+	p.mutPendingTx.Unlock()
+
+	if receivedAllMissing {
+		p.chReceivedAll <- true
+	}
+}
+
+func (p *pendingTransactions) getTransactionFromPool(txHash []byte) (data.TransactionHandler, bool) {
+	mb := p.mapHashes[string(txHash)]
+	storeId := process.ShardCacherIdentifier(mb.SenderShardID, mb.ReceiverShardID)
+	shardTxStore := p.txPools[mb.Type].ShardDataStore(storeId)
+
+	val, ok := shardTxStore.Peek(txHash)
+	if !ok {
+		return nil, false
+	}
+
+	tx, ok := val.(data.TransactionHandler)
+	if !ok {
+		return nil, false
+	}
+
+	return tx, true
+}
+
+func (p *pendingTransactions) getTransactionFromPoolOrStorage(hash []byte) (data.TransactionHandler, bool) {
+	txFromPool, ok := p.getTransactionFromPool(hash)
+	if ok {
+		return txFromPool, true
+	}
+
+	miniBlock, ok := p.mapHashes[string(hash)]
+	if !ok {
+		return nil, false
+	}
+
+	txData, err := GetDataFromStorage(hash, p.storage[miniBlock.Type], p.epochToSync)
+	if err != nil {
+		return nil, false
+	}
+
+	var tx data.TransactionHandler
```

Comment:
A function like base.getTransactionKeyTypeAndHash whould have been great

Reply:
key is not needed here. Syncer does not know about export keys and import keys.